core/src/main/scala/cats/collections/TreeList.scala

```diff
@@ -0,0 +1,867 @@
+      l.uncons match {
+        case None => ()
+        case Some((h, t)) =>
+          if (!first) strb.append(", ")
+          strb.append(h.toString)
+          loop(false, t)
+      }
+
+    loop(true, this)
+    strb.append(")")
+    strb.toString
+  }
+
+  /**
+   * Concatenate two TreeLists. This requires doing as
+   * much work as this.size
+   */
+  final def ++[A1 >: A](that: TreeList[A1]): TreeList[A1] = {
+    @tailrec
+    def loop(ls: List[A], that: TreeList[A1]): TreeList[A1] =
+      ls match {
+        case Nil => that
+        case h :: tail => loop(tail, h :: that)
+      }
+    loop(toListReverse, that)
+  }
+
+  /**
+   * keep the elements that match a predicate
+   */
+  final def filter(fn: A => Boolean): TreeList[A] = {
+    val as = toIterator
+    var resList = List.empty[A]
+    var changed = false
+    while(as.hasNext) {
+      val a = as.next()
+      if (fn(a)) {
+        resList = a :: resList
+      } else { changed = true }
+    }
+    if (changed) TreeList.fromListReverse(resList)
+    else this
+  }
+
+  final def filterNot(fn: A => Boolean): TreeList[A] = {
+    val as = toIterator
+    var resList = List.empty[A]
+    var changed = false
+    while(as.hasNext) {
+      val a = as.next()
+      if (!fn(a)) {
+        resList = a :: resList
+      } else { changed = true }
+    }
+    if (changed) TreeList.fromListReverse(resList)
+    else this
+  }
+
+  /**
+   * Standard flatMap on a List type.
+   */
+  final def flatMap[B](fn: A => TreeList[B]): TreeList[B] = {
+    @tailrec
+    def loop(rev: List[A], acc: TreeList[B]): TreeList[B] =
+      rev match {
+        case Nil => acc
+        case h :: tail =>
+          loop(tail, fn(h) ++ acc)
+      }
+    loop(toListReverse, TreeList.Empty)
+  }
+
+  /**
+   * O(N) reversal of the treeList
+   */
+  final def reverse: TreeList[A] = {
+    val revTrees = toIterator
+    var res: TreeList[A] = TreeList.Empty
+    while(revTrees.hasNext) {
+      res = revTrees.next() :: res
+    }
+    res
+  }
+
+  /**
+   * Take the first n things off the list. O(n)
+   */
+  final def take(n: Long): TreeList[A] = {
+    val takeIt = toIterator
+    var cnt = 0L
+    var res = List.empty[A]
+    while(takeIt.hasNext && cnt < n) {
+      res = takeIt.next() :: res
+      cnt += 1L
+    }
+    TreeList.fromListReverse(res)
+  }
+
+  /**
+   * If the given index is in the list, update and return Some(updated).
+   * else return None
+   *
+   * O(log N)
+   */
+  final def updated[A1 >: A](idx: Long, value: A1): Option[TreeList[A1]] = {
+    val up = updatedOrThis(idx, value)
+    if (up eq this) None else Some(up)
+  }
+}
+
+object TreeList extends TreeListInstances0 {
+  private object Impl {
+    sealed trait Nat {
+      def value: Int
+    }
+    object Nat {
+      case class Succ[P <: Nat](prev: P) extends Nat {
+        val value: Int = prev.value + 1
+      }
+      case object Zero extends Nat {
+        def value: Int = 0
```

Comment:
since it is returning a constant, it does not need to be. in fact, I think it only makes the object take more space if it does.

The Succ case needs a val otherwise it is O(value) cost to compute value.